### ISSUE TITLE
Promtool should support openmetrics

### ISSUE THREAD

**Issue by roidelapluie** [prometheus/prometheus#8932]
Promtool should support openmetrics, next to the text format.

Maybe via --format=openmetrics

**Comment by ide-rea** [comment]
Hi, i'm not sure does this mean check openmetrics format like check prom text format?

**Comment by roidelapluie** [comment]
This issue is to add support for `promtool check metrics --format=openmetrics`

**Comment by ide-rea** [comment]
Thanks, i would like to try  if not anybody is working on this.

**Comment by ide-rea** [comment]
Hi, i'm a little confused about the check rule. Do we parse the text wire openmetrics format into  prometheus MetricFamily, and then encoded the metrics into bytes stream, at last use promlint package to do the check?  need you help :) @roidelapluie

**Comment by roidelapluie** [comment]
This issue would require a openmetrics text parser in go, which does not exist at the moment. We only have a textformat parser in go.

**Comment by roidelapluie** [comment]
Here is the textparser in go: https://github.com/prometheus/common/blob/main/expfmt/text_parse.go
We should have one openmetrics parser in go for this issue. It should be developed next to the text parser, in the prometheus/common repository.

**Comment by ide-rea** [comment]
Thanks, i'm understanding the issue better!

**Comment by jyz0309** [comment]
Is this Issue still valid and can I try it?

**Comment by machine424** [comment]
I think so, especially that an openmetrics parser is now available https://github.com/prometheus/prometheus/blob/v2.47.0/model/textparse/openmetricsparse.go

**Comment by beorn7** [comment]
Hello from the bug scrub!

@jyz0309 are you still willing to work on this? We'll happily assign this issue to you (or to others that want to help with it).

Note that it requires an implementation of an OpenMetrics parsers in https://github.com/prometheus/common/tree/main/expfmt , so it's a bit of work and maybe not as easy as it looks.

**Comment by jyz0309** [comment]
@beorn7 Yes, I am willing to try this issue, I think I need to implement an OpenMetrics parser and merge in [expfmt](https://github.com/prometheus/common/blob/main/expfmt) firstly then use it to support openmetrics in promtool.

**Comment by jyz0309** [comment]
@beorn7 Hello! I want to ask sth about the parser in `expfmt`, I am implementing the OpenMetrics parser in `expfmt` and I found that in [client_model](https://github.com/prometheus/client_model/blob/10b17bce397d9241b0c31ebe461f782335bbf07f/go/metrics.pb.go#L41) still have not support `info, stateset` types, should I add these to protobuf file first or just unsupported these two type firstly (I found that [the encoder also unsupported these type yet](https://github.com/prometheus/common/blob/8742f090d9fd2cea7e53a4fd58ae5fb6c6634336/expfmt/openmetrics_create.go#L112))?

**Comment by jyz0309** [comment]
https://github.com/prometheus/common/pull/669
@beorn7 I have implement OpenMetrics parser in `expfmt` in this PR, PTAL, thanks :)

**Comment by martincostello** [comment]
I discovered I had a need for this while working on stabilising the OpenTelemetry .NET SDK's Prometheus exporter.

I've picked up the prior attempt and opened https://github.com/prometheus/common/pull/904 based on it.
